compiler/dyno/lib/uast/chpl-syntax-printer.cpp

```diff
@@ -0,0 +1,1065 @@
+    }
+
+    void visit(const VarArgFormal* node) {
+      if (node->intent() != Formal::Intent::DEFAULT_INTENT){
+        ss_ << kindToString(node->intent()) << " ";
+      }
+
+      ss_ << node->name().c_str();
+      if (const Expression* te = node->typeExpression()) {
+        ss_ << ": ";
+        printAst(ss_, te);
+      }
+      if (const Expression* ie = node->initExpression()) {
+        ss_ << " = ";
+        printAst(ss_, ie);
+      }
+
+      if (const Expression* ce = node->count()) {
+        ss_ << " ...";
+        if (const TypeQuery* tq = ce->toTypeQuery()){
+          printAst(ss_, tq);
+        } else {
+          printAst(ss_, ce);
+        }
+      }
+    }
+
+    void visit(const Variable* node) {
+      if (node->isConfig()) {
+        ss_ << "config ";
+      } else {
+        printLinkages(node);
+      }
+
+      if (node->kind() != Variable::Kind::INDEX) {
+        ss_ << kindToString(node->kind()) << " ";
+      }
+      ss_ << node->name().c_str();
+      if (const Expression* te = node->typeExpression()) {
+        ss_ << ": ";
+        printAst(ss_, te);
+      }
+      if (const Expression* ie = node->initExpression()) {
+        ss_ << " = ";
+        printAst(ss_, ie);
+      }
+    }
+
+    void visit(const VisibilityClause* node) {
+      if (node->limitationKind() == VisibilityClause::LimitationKind::ONLY &&
+          node->limitationKind() == VisibilityClause::LimitationKind::EXCEPT) {
+        ss_ << kindToString(node->limitationKind());
+        printAst(ss_, node->symbol());
+        interpose(ss_, node->limitations(), ", ");
+      } else if (node->limitationKind() == VisibilityClause::LimitationKind::BRACES) {
+        printAst(ss_, node->symbol());
+        interpose(ss_, node->limitations(), ", ", "{","}");
+      } else { //NONE
+        printAst(ss_, node->symbol());
+        interpose(ss_, node->limitations(), ", ");
+      }
+    }
+
+    void visit(const When* node) {
+      if (node->isOtherwise()) {
+        ss_ << "otherwise ";
+      } else {
+        ss_ << "when ";
+        interpose(ss_, node->caseExprs(), ", ");
+        ss_ << " ";
+      }
+      printBlockWithStyle(ss_, node->blockStyle(), node->stmts(), "do ");
+    }
+
+    void visit(const While* node) {
+      ss_ << "while ";
+      printAst(ss_, node->condition());
+      ss_ << " ";
+      printBlockWithStyle(ss_, node->blockStyle(), node->stmts(), "do ");
+    }
+
+    void visit(const WithClause* node) {
+      ss_ << "with ";
+      interpose(ss_, node->exprs(), ", ", "(", ")");
+    }
+
+    void visit(const Yield* node) {
+      ss_ << "yield ";
+      printAst(ss_, node->value());
+    }
+
+    void visit(const Zip* node) {
+      ss_ << "zip";
+      interpose(ss_, node->actuals(),", ","(",")");
+    }
+
+  };
+
+  /***************************************************************
+   * Specializations for generating userString when converting
+   * from uAST to old AST
+   */
+
+  struct UserStringVisitor {
```

Comment:
Style: I don't think we need a whole other struct here. For the most part it seems like all the work is done in the `ChplSyntaxVisitor`, we just need to copy the body of `visit(uast::Function*)` into a helper method and adjust it so we know when to stop (what you've done in the visitor below).